### DISCUSSION TITLE
Magick not reading color palette of grayscale tiff correctly.

### DISCUSSION THREAD

**Discussion by MrAnderson1971** [ImageMagick/ImageMagick#6405] [Help] [answered]
### ImageMagick version

7.1.1-Q8

### Operating system, version and so on

Windows 10 (x86)

### Description

[DFIRLEFT.zip](https://github.com/ImageMagick/ImageMagick/files/11728483/DFIRLEFT.zip)
Got this 8 bit grayscale TIFF file here. When I read it using ImageMagick and read its color palette, I get this:
0:#000000FF 1:#010101FF 2:#020202FF 3:#030303FF 4:#040404FF 5:#050505FF 6:#060606FF 7:#070707FF 8:#080808FF 9:#090909FF 10:#0A0A0AFF 11:#0B0B0BFF 12:#0C0C0CFF 13:#0D0D0DFF 14:#0E0E0EFF 15:#0F0F0FFF 16:#101010FF 17:#111111FF 18:#121212FF 19:#131313FF 20:#141414FF 21:#151515FF 22:#161616FF 23:#171717FF 24:#181818FF 25:#191919FF 26:#1A1A1AFF 27:#1B1B1BFF 28:#1C1C1CFF 29:#1D1D1DFF 30:#1E1E1EFF 31:#1F1F1FFF 32:#202020FF 33:#212121FF 34:#222222FF 35:#232323FF 36:#242424FF 37:#252525FF 38:#262626FF 39:#272727FF 40:#282828FF 41:#292929FF 42:#2A2A2AFF 43:#2B2B2BFF 44:#2C2C2CFF 45:#2D2D2DFF 46:#2E2E2EFF 47:#2F2F2FFF 48:#303030FF 49:#313131FF 50:#323232FF 51:#333333FF 52:#343434FF 53:#353535FF 54:#363636FF 55:#373737FF 56:#383838FF 57:#393939FF 58:#3A3A3AFF 59:#3B3B3BFF 60:#3C3C3CFF 61:#3D3D3DFF 62:#3E3E3EFF 63:#3F3F3FFF 64:#404040FF 65:#414141FF 66:#424242FF 67:#434343FF 68:#444444FF 69:#454545FF 70:#464646FF 71:#474747FF 72:#484848FF 73:#494949FF 74:#4A4A4AFF 75:#4B4B4BFF 76:#4C4C4CFF 77:#4D4D4DFF 78:#4E4E4EFF 79:#4F4F4FFF 80:#505050FF 81:#515151FF 82:#525252FF 83:#535353FF 84:#545454FF 85:#555555FF 86:#565656FF 87:#575757FF 88:#585858FF 89:#595959FF 90:#5A5A5AFF 91:#5B5B5BFF 92:#5C5C5CFF 93:#5D5D5DFF 94:#5E5E5EFF 95:#5F5F5FFF 96:#606060FF 97:#616161FF 98:#626262FF 99:#636363FF 100:#646464FF 101:#656565FF 102:#666666FF 103:#676767FF 104:#686868FF 105:#696969FF 106:#6A6A6AFF 107:#6B6B6BFF 108:#6C6C6CFF 109:#6D6D6DFF 110:#6E6E6EFF 111:#6F6F6FFF 112:#707070FF 113:#717171FF 114:#727272FF 115:#737373FF 116:#747474FF 117:#757575FF 118:#767676FF 119:#777777FF 120:#787878FF 121:#797979FF 122:#7A7A7AFF 123:#7B7B7BFF 124:#7C7C7CFF 125:#7D7D7DFF 126:#7E7E7EFF 127:#7F7F7FFF **128:#808080FF 129:#808080FF** 130:#818181FF 131:#828282FF 132:#838383FF 133:#848484FF 134:#858585FF 135:#868686FF 136:#878787FF 137:#888888FF 138:#898989FF 139:#8A8A8AFF 140:#8B8B8BFF 141:#8C8C8CFF 142:#8D8D8DFF 143:#8E8E8EFF 144:#8F8F8FFF 145:#909090FF 146:#919191FF 147:#929292FF 148:#939393FF 149:#949494FF 150:#959595FF 151:#969696FF 152:#979797FF 153:#989898FF 154:#999999FF 155:#9A9A9AFF 156:#9B9B9BFF 157:#9C9C9CFF 158:#9D9D9DFF 159:#9E9E9EFF 160:#9F9F9FFF 161:#A0A0A0FF 162:#A1A1A1FF 163:#A2A2A2FF 164:#A3A3A3FF 165:#A4A4A4FF 166:#A5A5A5FF 167:#A6A6A6FF 168:#A7A7A7FF 169:#A8A8A8FF 170:#A9A9A9FF 171:#AAAAAAFF 172:#ABABABFF 173:#ACACACFF 174:#ADADADFF 175:#AEAEAEFF 176:#AFAFAFFF 177:#B0B0B0FF 178:#B1B1B1FF 179:#B2B2B2FF 180:#B3B3B3FF 181:#B4B4B4FF 182:#B5B5B5FF 183:#B6B6B6FF 184:#B7B7B7FF 185:#B8B8B8FF 186:#B9B9B9FF 187:#BABABAFF 188:#BBBBBBFF 189:#BCBCBCFF 190:#BDBDBDFF 191:#BEBEBEFF 192:#BFBFBFFF 193:#C0C0C0FF 194:#C1C1C1FF 195:#C2C2C2FF 196:#C3C3C3FF 197:#C4C4C4FF 198:#C5C5C5FF 199:#C6C6C6FF 200:#C7C7C7FF 201:#C8C8C8FF 202:#C9C9C9FF 203:#CACACAFF 204:#CBCBCBFF 205:#CCCCCCFF 206:#CDCDCDFF 207:#CECECEFF 208:#CFCFCFFF 209:#D0D0D0FF 210:#D1D1D1FF 211:#D2D2D2FF 212:#D3D3D3FF 213:#D4D4D4FF 214:#D5D5D5FF 215:#D6D6D6FF 216:#D7D7D7FF 217:#D8D8D8FF 218:#D9D9D9FF 219:#DADADAFF 220:#DBDBDBFF 221:#DCDCDCFF 222:#DDDDDDFF 223:#DEDEDEFF 224:#DFDFDFFF 225:#E0E0E0FF 226:#E1E1E1FF 227:#E2E2E2FF 228:#E3E3E3FF 229:#E4E4E4FF 230:#E5E5E5FF 231:#E6E6E6FF 232:#E7E7E7FF 233:#E8E8E8FF 234:#E9E9E9FF 235:#EAEAEAFF 236:#EBEBEBFF 237:#ECECECFF 238:#EDEDEDFF 239:#EEEEEEFF 240:#EFEFEFFF 241:#F0F0F0FF 242:#F1F1F1FF 243:#F2F2F2FF 244:#F3F3F3FF 245:#F4F4F4FF 246:#F5F5F5FF 247:#F6F6F6FF 248:#F7F7F7FF 249:#F8F8F8FF 250:#F9F9F9FF 251:#FAFAFAFF 252:#FBFBFBFF 253:#FCFCFCFF 254:#FDFDFDFF 255:#FEFEFEFF 

The 0x808080 entry appears twice, and the 0xFFFFFF is missing. Is there any way to make sure all values from 0x000000 to 0xFFFFFF appear once in order? 

Here is the code:
`Magick::Image image;`

`image.read(file);`

`for (size_t i = 0; i < image.colorMapSize(); i++) {`

    `std::cout << i << ":" << std::string(image.colorMap(i)) << " ";`

`}`

**Comment by urban-warrior** [comment]
Try `identify -verbose DFIRLEFT.TIF`, you'll find the image colormap does not include perfect white.

**Reply by MrAnderson1971** [reply]
118: (118,118,118,1) #767676FF srgba(118,118,118,1)
    119: (119,119,119,1) #777777FF srgba(119,119,119,1)
    120: (120,120,120,1) #787878FF grey47
    121: (121,121,121,1) #797979FF srgba(121,121,121,1)
    122: (122,122,122,1) #7A7A7AFF grey48
    123: (123,123,123,1) #7B7B7BFF srgba(123,123,123,1)
    124: (124,124,124,1) #7C7C7CFF srgba(124,124,124,1)
    125: (125,125,125,1) #7D7D7DFF grey49
    126: (126,126,126,1) #7E7E7EFF gray
    127: (127,127,127,1) #7F7F7FFF grey50
    128: (128,128,128,1) #808080FF fractal
    129: (128,128,128,1) #808080FF fractal
    130: (129,129,129,1) #818181FF srgba(129,129,129,1)
    131: (130,130,130,1) #828282FF grey51
    132: (131,131,131,1) #838383FF srgba(131,131,131,1)
    133: (132,132,132,1) #848484FF srgba(132,132,132,1)
    134: (133,133,133,1) #858585FF grey52
    135: (134,134,134,1) #868686FF srgba(134,134,134,1)

"fractal"?

**Reply by fmw42** [reply]
fractal is just a named color for mid gray.  

`fractal  rgb(128, 128, 128)  #808080`

See https://imagemagick.org/script/color.php

**Reply by MrAnderson1971** [reply]
However, in the actual palette inside of the tiff file, 0x80 only appears once, and there is an 0xFF as well, which is different from what Magick reads:
![MicrosoftTeams-image](https://github.com/ImageMagick/ImageMagick/assets/67478505/206aebed-b71c-4c91-88d9-96597df36fba)

**Reply by fmw42** [reply]
Post your tiff file. Zip it first before uploading or post the zip file to some free hosting service and put the URL here if the file is too big to upload.

**Reply by MrAnderson1971** [reply]
[DFIRLEFT.zip](https://github.com/ImageMagick/ImageMagick/files/11737664/DFIRLEFT.zip)

**Comment by snibgo** [comment]
`#808080FF` is a colour rounded to 8 bits per channel. But the palette entries have 16 bits per channel. Hence two 16-bit values round to the same 8-bit value.

**Comment by snibgo** [comment, marked as answer]
To expand on my previous post, we can examine the values of the unique colours:
```
magick c:\web\im\DFIRLEFT.TIF -unique-colors -sort-pixels -precision 16 -define ftxt:format="\v \h \H\n" ftxt:
```

The output includes:
```
0,0,0 #0,#0,#0 #000000
256,256,256 #100,#100,#100 #010101
:
32256,32256,32256 #7e00,#7e00,#7e00 #7E7E7E
32512,32512,32512 #7f00,#7f00,#7f00 #7F7F7F
32768,32768,32768 #8000,#8000,#8000 #808080
33024,33024,33024 #8100,#8100,#8100 #808080
33280,33280,33280 #8200,#8200,#8200 #818181
33536,33536,33536 #8300,#8300,#8300 #828282
:
65024,65024,65024 #fe00,#fe00,#fe00 #FDFDFD
65280,65280,65280 #ff00,#ff00,#ff00 #FEFEFE
```
This shows the RGB values in three formats. First is decimal, in the range 0 to QuantumRange, which here (Q16) is 65535. The image is grayscale, so the three channel values are equal.

The second format shows each channel as a hex number. For example, 32768 is #8000, and 33024 is #8100.

The third format shows the channels rounded to 8 bits (two hex digits), and the three channels are concatenated. #8000 rounded to 8 bits is #80. #8100 rounded to 8 bits is also #80. 

Hence two different 16-bit values are the same when rounded to 8 bits.
